File: src/lib/landscape.ts
import yaml from 'js-yaml';
import type { LandscapeData, LandscapeProject } from './schemas';

const LANDSCAPE_URL = 'https://raw.githubusercontent.com/cncf/landscape/master/landscape.yml';

/**
 * Fetch and parse CNCF landscape.yml
 * Returns a map of org/repo -> project metadata
 */
export async function fetchLandscapeData(): Promise<LandscapeData> {
  console.log('[Landscape] Fetching from:', LANDSCAPE_URL);
  
  try {
    const response = await fetch(LANDSCAPE_URL);
    if (!response.ok) {
      throw new Error(`Failed to fetch landscape: ${response.status} ${response.statusText}`);
    }
    
    const yamlText = await response.text();
    console.log('[Landscape] Downloaded', yamlText.length, 'bytes');
    
    const parsed = yaml.load(yamlText) as any;
    const projectMap = parseLandscapeYaml(parsed);
    
    console.log('[Landscape] Parsed', Object.keys(projectMap).length, 'projects');
    return projectMap;
  } catch (error) {
    console.error('[Landscape] Error fetching landscape:', error);
    throw error;
  }
}

/**
 * Parse landscape YAML structure into flat project map
 * Structure: landscape -> categories -> subcategories -> items
 */
function parseLandscapeYaml(data: any): LandscapeData {
  const projectMap: LandscapeData = {};
  
  if (!data.landscape || !Array.isArray(data.landscape)) {
    console.warn('[Landscape] No landscape array found in data');
    return projectMap;
  }
  
  // Iterate through categories
  for (const category of data.landscape) {
    if (!category.subcategories || !Array.isArray(category.subcategories)) {
      continue;
    }
    
    // Iterate through subcategories
    for (const subcategory of category.subcategories) {
      if (!subcategory.items || !Array.isArray(subcategory.items)) {
        continue;
      }
      
      // Iterate through items (projects)
      for (const item of subcategory.items) {
        const project = parseProjectItem(item);
        if (project && project.repo_url) {
          // Extract org/repo from GitHub URL
          const slug = extractRepoSlug(project.repo_url);
          if (slug) {
            // Only overwrite if new project has CNCF status or existing doesn't
            const existing = projectMap[slug];
            if (!existing || project.project || !existing.project) {
              projectMap[slug] = project;
            }
            // If both have status, prefer the CNCF one (non-Wasm, non-duplicate)
            if (existing && existing.project && project.project) {
              // Keep the one without "(Wasm)" or other suffixes in the name
              if (!project.name.includes('(') && existing.name.includes('(')) {
                projectMap[slug] = project;
              }
            }
          }
        }
      }
    }
  }
  
  return projectMap;
}

/**
 * Parse individual project item from landscape
 */
function parseProjectItem(item: any): LandscapeProject | null {
  if (!item || typeof item !== 'object') {
    return null;
  }
  
  // Extract project status from 'project' field
  let projectStatus: 'graduated' | 'incubating' | 'sandbox' | undefined;
  if (item.project === 'graduated' || item.project === 'incubating' || item.project === 'sandbox') {
    projectStatus = item.project;
  }
  
  // Extract description from extra.summary_use_case or extra.summary_business_use_case
  let description: string | undefined;
  if (item.extra) {
    description = item.extra.summary_use_case || item.extra.summary_business_use_case;
    // Clean up description (remove extra whitespace)
    if (description) {
      description = description.replace(/\s+/g, ' ').trim();
    }
  }
  
  return {
    name: item.name || '',
    description,
    repo_url: item.repo_url,
    homepage_url: item.homepage_url,
    project: projectStatus,
  };
}

/**
 * Extract org/repo slug from GitHub URL
 * Example: https://github.com/dapr/dapr -> dapr/dapr
 */
export function extractRepoSlug(url: string): string | null {
  try {
    const parsed = new URL(url);
    if (parsed.hostname !== 'github.com') {
      return null;
    }
    
    // Remove leading/trailing slashes and extract org/repo
    const parts = parsed.pathname.replace(/^\/|\/$/g, '').split('/');
    if (parts.length >= 2) {
      return `${parts[0]}/${parts[1]}`;
    }
    
    return null;
  } catch {
    return null;
  }
}

/**
 * Match a feed URL to a landscape project
 * Example: https://github.com/dapr/dapr/releases.atom -> dapr/dapr
 */
export function matchFeedToProject(feedUrl: string, landscapeData: LandscapeData): LandscapeProject | null {
  const slug = extractRepoSlug(feedUrl);
  if (!slug) {
    return null;
  }
  
  return landscapeData[slug] || null;
}

File: firehose-go/internal/landscape/landscape.go
package landscape

import (
	"fmt"
	"io"
	"net/http"

	"github.com/castrojo/firehose-go/internal/models"
	"gopkg.in/yaml.v3"
)

const landscapeURL = "https://raw.githubusercontent.com/cncf/landscape/master/landscape.yml"

// FetchAndParse fetches and parses the CNCF Landscape
func FetchAndParse() (map[string]models.LandscapeProject, error) {
	// Fetch landscape.yml
	resp, err := http.Get(landscapeURL)
	if err != nil {
		return nil, fmt.Errorf("fetch landscape: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("fetch landscape: HTTP %d", resp.StatusCode)
	}

	// Read response body
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read landscape: %w", err)
	}

	// Parse YAML
	return parseLandscapeYAML(data)
}

// parseLandscapeYAML parses landscape.yml and builds org/repo → project map
func parseLandscapeYAML(data []byte) (map[string]models.LandscapeProject, error) {
	var landscape struct {
		Landscape []struct {
			Name       string `yaml:"name"`
			Categories []struct {
				Name          string `yaml:"name"`
				Subcategories []struct {
					Name  string `yaml:"name"`
					Items []struct {
						Name           string `yaml:"name"`
						Homepage       string `yaml:"homepage_url"`
						Repo           string `yaml:"repo_url"`
						Project        string `yaml:"project"` // graduated, incubating, sandbox
						SummaryUseCase string `yaml:"summary_use_case"`
						Description    string `yaml:"description"`
					} `yaml:"items"`
				} `yaml:"subcategories"`
			} `yaml:"categories"`
		} `yaml:"landscape"`
	}

	if err := yaml.Unmarshal(data, &landscape); err != nil {
		return nil, fmt.Errorf("parse YAML: %w", err)
	}

	// Build map keyed by org/repo
	projectMap := make(map[string]models.LandscapeProject)

	for _, section := range landscape.Landscape {
		for _, category := range section.Categories {
			for _, subcategory := range category.Subcategories {
				for _, item := range subcategory.Items {
					if item.Repo == "" {
						continue
					}

					// Extract org/repo from GitHub URL
					// Example: https://github.com/kubernetes/kubernetes → kubernetes/kubernetes
					slug := extractOrgRepo(item.Repo)
					if slug == "" {
						continue
					}

					description := item.Description
					if description == "" {
						description = item.SummaryUseCase
					}

					projectMap[slug] = models.LandscapeProject{
						Name:        item.Name,
						Description: description,
						RepoURL:     item.Repo,
						HomepageURL: item.Homepage,
						Status:      item.Project,
					}
				}
			}
		}
	}

	return projectMap, nil
}

// extractOrgRepo extracts org/repo from GitHub URL
// Example: https://github.com/kubernetes/kubernetes → kubernetes/kubernetes
func extractOrgRepo(repoURL string) string {
	// Simple extraction: find "github.com/" and take next two path segments
	const githubPrefix = "github.com/"
	idx := -1

	// Find github.com in URL (handles both http and https)
	for i := range repoURL {
		if i+len(githubPrefix) <= len(repoURL) && repoURL[i:i+len(githubPrefix)] == githubPrefix {
			idx = i + len(githubPrefix)
			break
		}
	}

	if idx == -1 {
		return ""
	}

	// Extract org/repo (everything after github.com/)
	remainder := repoURL[idx:]

	// Find first and second slash
	firstSlash := -1
	secondSlash := -1
	for i, c := range remainder {
		if c == '/' {
			if firstSlash == -1 {
				firstSlash = i
			} else if secondSlash == -1 {
				secondSlash = i
				break
			}
		}
	}

	if firstSlash == -1 {
		return ""
	}

	// If no second slash, return everything (e.g., "kubernetes/kubernetes")
	if secondSlash == -1 {
		return remainder
	}

	// Return org/repo (e.g., "kubernetes/kubernetes" from "kubernetes/kubernetes/releases")
	return remainder[:secondSlash]
}
